drivers/gpu/drm/vc4/vc4_crtc.c

```diff
@@ -377,31 +381,38 @@ static void vc4_crtc_config_pv(struct drm_crtc *crtc)
 		   VC4_SET_FIELD(mode->crtc_vdisplay, PV_VERTB_VACTIVE));
 
 	if (interlace) {
+		u32 field_delay = mode->htotal * pixel_rep / 2;
+		if (vec_interlaced)
+			++field_delay;
+
 		CRTC_WRITE(PV_VERTA_EVEN,
-			   VC4_SET_FIELD(mode->crtc_vtotal -
-					 mode->crtc_vsync_end - 1,
+			   VC4_SET_FIELD(mode->crtc_vtotal + vtotal_fixup -
+					 mode->crtc_vsync_end -
+					 (ntsc ? 0 : 1),
 					 PV_VERTA_VBP) |
 			   VC4_SET_FIELD(mode->crtc_vsync_end -
 					 mode->crtc_vsync_start,
 					 PV_VERTA_VSYNC));
 		CRTC_WRITE(PV_VERTB_EVEN,
 			   VC4_SET_FIELD(mode->crtc_vsync_start -
-					 mode->crtc_vdisplay,
+					 mode->crtc_vdisplay +
+					 (ntsc ? 1 : 0),
 					 PV_VERTB_VFP) |
-			   VC4_SET_FIELD(mode->crtc_vdisplay, PV_VERTB_VACTIVE));
+			   VC4_SET_FIELD(mode->crtc_vdisplay,
+					 PV_VERTB_VACTIVE));
```

Comment:
All of these fixups become quite hard to read and understand what exactly is going on.

My gut feeling is that it'd be clearer to add local variables at least for the vertical active, fp, sw, and bp values, and then have explicit clauses for NTSC and PAL when assigning to the registers. At least that way the fixups are obvious.

Reply:
OK, I'll try to simplify things a bit.